File: object-batch-connection/ios/SQLiteBatchCore.m
#import "SQLiteBatchCore.h"

#include "sqlite-connection-core.h"

@implementation SQLiteBatchCore

+ (void) initialize
{
  scc_init();
}

+ (void) openDatabaseConnection: (NSArray *) arguments
                        success: (void (^)(int)) successCallback
                          error: (void (^)(NSString *)) errorCallback
{
  NSArray * _args = arguments;

  NSDictionary * options = (NSDictionary *)[_args objectAtIndex: 0];

  const char * filename = [(NSString *)[options valueForKey: @"path"] cString];

  const int flags = [(NSNumber *)[options valueForKey: @"flags"] intValue];

  const int connection_id = scc_open_connection(filename, flags);

  if (connection_id < 0) {
    errorCallback( @"open error");
    return;
  }

  successCallback(connection_id);
}

+ (void) executeBatch: (NSArray *) arguments
              success: (void (^)(NSArray *)) successCallback
{
  NSArray * _args = arguments;

  const int connection_id = [(NSNumber *)[_args objectAtIndex: 0] intValue];

  NSArray * data = [_args objectAtIndex: 1];

  NSMutableArray * results = [NSMutableArray arrayWithCapacity: 0];

  for (int i=0; i < [data count]; ++i) {
    int previousTotalChanges = scc_get_total_changes(connection_id);

    NSArray * entry = [data objectAtIndex:i];

    NSString * statement = [entry objectAtIndex: 0];

    NSArray * bind = [entry objectAtIndex: 1];

    int prepareResult =
      scc_begin_statement(connection_id, [statement cString]);

    if (prepareResult == 0) {
      for (int j=0; j < [bind count]; ++j) {
        NSObject * bindValue = [bind objectAtIndex: j];

        if (bindValue == nil) {
          prepareResult = scc_bind_null(connection_id, 1 + j);
        } else if ([bindValue isKindOfClass: [NSNumber class]]) {
          // TBD UIWebView vs WKWebView
          if ([(NSNumber *)bindValue objCType][0] == 'q') {
            prepareResult = scc_bind_long(connection_id, 1 + j,
              [(NSNumber *)bindValue longValue]);
          } else {
            prepareResult = scc_bind_double(connection_id, 1 + j,
              [(NSNumber *)bindValue doubleValue]);
          }
        } else if ([bindValue isKindOfClass: [NSString class]]) {
          prepareResult = scc_bind_text(connection_id, 1 + j,
              [(NSString *)bindValue cString]);
        } else {
          prepareResult = scc_bind_null(connection_id, 1 + j);
        }
      }
    }

    if (prepareResult != 0) {
      [results addObject: @{
        @"status": @1,
        @"message": [NSString stringWithUTF8String:
          scc_get_last_error_message(connection_id)]
      }];

      scc_end_statement(connection_id);

      continue;
    }

    int stepResult = scc_step(connection_id);

    if (stepResult == 100) {
      const int columnCount = scc_get_column_count(connection_id);

      NSMutableArray * columns = [NSMutableArray arrayWithCapacity: 0];

      for (int j = 0; j < columnCount; ++j) {
        NSString * columnNameAsString =
          [NSString stringWithUTF8String: scc_get_column_name(connection_id, j)];
        [columns addObject: columnNameAsString];
      }

      NSMutableArray * rows = [NSMutableArray arrayWithCapacity: 0];

      do {
        NSMutableArray * row = [NSMutableArray arrayWithCapacity: 0];

        for (int j = 0; j < columnCount; ++j) {
          const int columnType = scc_get_column_type(connection_id, j);

          if (columnType == SCC_COLUMN_TYPE_NULL) {
            [row addObject: [NSNull null]];
          } else if (columnType == SCC_COLUMN_TYPE_INTEGER) {
            NSNumber * columnNumberValue =
              [NSNumber numberWithLongLong: scc_get_column_long(connection_id, j)];
            [row addObject: columnNumberValue];
          } else if (columnType == SCC_COLUMN_TYPE_FLOAT) {
            NSNumber * columnNumberValue =
              [NSNumber numberWithDouble: scc_get_column_double(connection_id, j)];
            [row addObject: columnNumberValue];
          } else {
            NSString * columnStringValue =
              [NSString stringWithUTF8String: scc_get_column_text(connection_id, j)];
            [row addObject: columnStringValue];
          }
        }

        [rows addObject: row];

        stepResult = scc_step(connection_id);
      } while(stepResult == 100);

      [results addObject: @{@"status":@0, @"columns": columns, @"rows": rows}];
    } else if (stepResult == 101) {
      int totalChanges = scc_get_total_changes(connection_id);
      int rowsAffected = totalChanges - previousTotalChanges;

      [results addObject: @{
        @"status": @0,
        @"rowsAffected": [NSNumber numberWithInteger: rowsAffected],
        @"totalChanges": [NSNumber numberWithInteger: totalChanges],
        @"lastInsertRowId": [NSNumber numberWithInteger:
          scc_get_last_insert_rowid(connection_id)]
      }];
    } else {
      [results addObject: @{
        @"status": @1,
        @"message": [NSString stringWithUTF8String:
          scc_get_last_error_message(connection_id)]
      }];
    }

    scc_end_statement(connection_id);
  }

  successCallback(results);
}

@end

File: object-batch-connection/android/io/sqlc/SQLiteBatchCore.java
package io.sqlc;

import org.json.JSONArray;
import org.json.JSONObject;

public class SQLiteBatchCore {
  public interface OpenCallbacks {
    public void success(int connectionId);
    public void error(String message);
  }

  public interface ExecuteBatchCallbacks {
    public void success(JSONArray results);
    public void error(String message);
  }

  static public void
  openBatchConnection(String pathname, int flags, OpenCallbacks cb) {
    final int mydbc = SCCoreGlue.scc_open_connection(pathname, flags);

    if (mydbc < 0) {
      cb.error("open error: " + -mydbc);
    } else {
      cb.success(mydbc);
    }
  }

  static public void
  executeBatch(int mydbc, JSONArray data, ExecuteBatchCallbacks cb) {
    try {
      final int count = data.length();

      JSONArray results = new JSONArray();

      for (int i=0; i<count; ++i) {
        int previousTotalChanges = SCCoreGlue.scc_get_total_changes(mydbc);

        JSONArray entry = data.getJSONArray(i);

        String s = entry.getString(0);

        if (SCCoreGlue.scc_begin_statement(mydbc, s) != 0) {
          JSONObject result = new JSONObject();
          result.put("status", 1); // REPORT SQLite ERROR 1
          result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
          results.put(result);
        } else {
          JSONArray bind = entry.getJSONArray(1);

          final int bindCount = bind.length();

          int bindResult = 0; // SQLite OK

          for (int j = 0; j < bindCount; ++j) {
            final Object o = bind.get(j);

            if (o instanceof Number) {
              bindResult =
                SCCoreGlue.scc_bind_double(mydbc, 1 + j, bind.optDouble(j));
            } else if (o instanceof String) {
              bindResult =
                SCCoreGlue.scc_bind_text(mydbc, 1 + j, o.toString());
            } else {
              bindResult =
                SCCoreGlue.scc_bind_null(mydbc, 1 + j);
            }
          }

          if (bindResult != 0) {
            JSONObject result = new JSONObject();
            result.put("status", 1); // REPORT SQLite ERROR 1
            result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
            continue;
          }

          int stepResult = SCCoreGlue.scc_step(mydbc);

          if (stepResult == 100) {
            final int columnCount = SCCoreGlue.scc_get_column_count(mydbc);

            JSONArray columns = new JSONArray();

            for (int j=0; j < columnCount; ++j) {
              columns.put(SCCoreGlue.scc_get_column_name(mydbc, j));
            }

            JSONArray rows = new JSONArray();

            do {
              JSONArray row = new JSONArray();

              for (int col=0; col < columnCount; ++col) {
                final int type = SCCoreGlue.scc_get_column_type(mydbc, col);

                if (type == SCCoreGlue.SCC_COLUMN_TYPE_INTEGER) {
                  row.put(SCCoreGlue.scc_get_column_long(mydbc, col));
                } else if (type == SCCoreGlue.SCC_COLUMN_TYPE_FLOAT) {
                  row.put(SCCoreGlue.scc_get_column_double(mydbc, col));
                } else if (type == SCCoreGlue.SCC_COLUMN_TYPE_NULL) {
                  row.put(JSONObject.NULL);
                } else {
                  row.put(SCCoreGlue.scc_get_column_text(mydbc, col));
                }
              }

              rows.put(row);

              stepResult = SCCoreGlue.scc_step(mydbc);
            } while (stepResult == 100);

            JSONObject result = new JSONObject();
            result.put("status", 0); // REPORT SQLite OK
            result.put("columns", columns);
            result.put("rows", rows);
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          } else if (stepResult == 101) {
            int totalChanges = SCCoreGlue.scc_get_total_changes(mydbc);
            int rowsAffected = totalChanges - previousTotalChanges;

            JSONObject result = new JSONObject();
            // same order as iOS & macOS ("osx"):
            result.put("status", 0); // REPORT SQLite OK
            result.put("totalChanges", totalChanges);
            result.put("rowsAffected", rowsAffected);
            result.put("lastInsertRowId",
              SCCoreGlue.scc_get_last_insert_rowid(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          } else {
            JSONObject result = new JSONObject();
            result.put("status", 1); // REPORT SQLite ERROR 1
            result.put("message", SCCoreGlue.scc_get_last_error_message(mydbc));
            results.put(result);
            SCCoreGlue.scc_end_statement(mydbc);
          }
        }
      }

      cb.success(results);
    } catch(Exception e) {
      // NOT EXPECTED - internal error:
      cb.error(e.toString());
    }
  }

  static {
    System.loadLibrary("sqlite-connection-core-glue");
    SCCoreGlue.scc_init();
  }
}
